### PR TITLE
Fix osx mojave build

### PR DESCRIPTION
**Pull request by dm1try**
ref #9050 
cc @justinmk 

I decided to not put SDK path to each third party dep alongside with `CC` value; does it look dirty? 😄

### CODE REVIEW

File: third-party/CMakeLists.txt [hunk 1]
```diff
@@ -98,6 +98,10 @@ else()
   set(DEPS_C_COMPILER "${CMAKE_C_COMPILER}")
 endif()
 
+if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+  set(DEPS_C_COMPILER "${DEPS_C_COMPILER} -isysroot${CMAKE_OSX_SYSROOT}")
```
Comment (dm1try) [inline third-party/CMakeLists.txt]: should I restrict this for running on Mojave only? theoretically(!), this should work on old osx  too)

Comment (jamessan) [reply third-party/CMakeLists.txt]: Ideally it is generally applicable, but we'll find out since Travis is running an older macOS. :)

**Comment by jamessan** [comment]
Why do these changes only need to be made to the third-party CMakeLists.txt and not neovim's?

File: third-party/CMakeLists.txt (hunk 1 above)
Comment (dm1try) [reply third-party/CMakeLists.txt]: yep, the build is red :( `CMAKE_OSX_SYSROOT` is empty

**Comment by dm1try** [comment]
@jamessan, actually see this https://github.com/neovim/neovim/blob/8fd092f3ff15bf70f84ec0d716c5aaa2c7379fa1/src/nvim/CMakeLists.txt#L190-L193

File: third-party/cmake/BuildLuajit.cmake [hunk 2]
```diff
@@ -50,8 +50,15 @@ set(INSTALLCMD_UNIX ${MAKE_PRG} CFLAGS=-fPIC
                                 install)
 
 if(UNIX)
+  if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+    set(DEPLOYMENT_TARGET "MACOSX_DEPLOYMENT_TARGET=${CMAKE_OSX_DEPLOYMENT_TARGET}")
```
Comment (justinmk) [inline third-party/cmake/BuildLuajit.cmake]: add a comment:

```
# luajit defaults MACOSX_DEPLOYMENT_TARGET to 10.4
# https://github.com/LuaJIT/LuaJIT/blob/b025b01c5b9d23f6218c7d72b7aafa3f1ab1e08a/src/Makefile#L301-L303
```

**Comment by justinmk** [comment]
> CMAKE_OSX_SYSROOT is properly resolved, but the compiler of third party deps somehow does not use it;

In fact we do the same thing for the "generator" (#7891): https://github.com/neovim/neovim/blob/27d755cfd1a4a7f571871a1304184f8b009546b4/src/nvim/CMakeLists.txt#L190-L193

**Comment by dm1try** [comment]
> In fact we do the same thing for the "generator" build task (#7891):

yep, I saw #7891, though it does not have information about "why does it behave so?"
> It turns out there's a difference between executing /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang and /usr/bin/clang. CMake was picking up the former and with that it left off a key include path. Why?

`/usr/bin/clang` is [a shim that uses `xcrun` under the hood](https://macops.ca/developer-binaries-on-os-x-xcode-select-and-xcrun/).
```
otool -L /usr/bin/clang
/usr/bin/clang:
 /usr/lib/libxcselect.dylib (compatibility version 1.0.0, current version 1.0.0)
 /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
```
>this shim binary loads functions in libxcselect.dylib that can locate the path to the real binary, depending on how the system has been configured. One part of this process is to check whether this path contains usr/lib/libxcrun.dylib, and the xcrun tool, in which case it will invoke xcrun to run the binary.

 if you read [the first part](https://github.com/neovim/neovim/issues/9050#issuecomment-429674155) of this investigation 😩 you have more clarity now.

**Comment by dm1try** [comment]
Could you guys also help me to parse this line about `CMAKE_OSX_SYSROOT`
> CMake uses this value to compute the value of the -isysroot flag 

I read this like "Hey man if the CMAKE_OSX_SYSROOT is set I add `-isysroot` flag for you on compiling". Or it means we should provide the flag anyway but without a value?

**Comment by dm1try** [comment]
NOTE: after merging this, the [wiki build page](https://github.com/neovim/neovim/wiki/Building-Neovim#macos) should be updated

**Comment by justinmk** [comment]
> Could you guys also help me to parse [this line](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_SYSROOT.html) about `CMAKE_OSX_SYSROOT`
> 
> > CMake uses this value to compute the value of the -isysroot flag
> 
> I read this like "Hey man if CMAKE_OSX_SYSROOT is set I add `-isysroot` flag for you on compiling". Or it means we should provide the flag anyway but without a value?

Neither. It means CMake uses that to set `-isysroot`, otherwise it tries to figure out `-isysroot` automatically.

> NOTE: after merging this, the wiki build page should be updated

Feel free to update it directly.

**Pull request merged**